File: Data Structures & Algorithms/matrixDFS/submission-1.java
public class Solution {

    public int countPaths(int[][] grid) {
        Set<String> visited = new HashSet<>();
        return countPaths(grid, visited, 0, 0);
    }

    public int countPaths(int[][] grid, Set<String> visited, int row, int col) {
        if (visited.contains(row + ":" + col) || !isInRange(grid, row, col) || grid[row][col] == 1) {
            return 0;
        }
        if (row == grid.length - 1 && col == grid[0].length - 1) {
            return 1;
        }
        int res = 0;
        String key = row + ":" + col;
        visited.add(key);
        res += countPaths(grid, visited, row, col + 1);
        res += countPaths(grid, visited, row + 1, col);
        res += countPaths(grid, visited, row, col - 1);
        res += countPaths(grid, visited, row - 1, col);
        visited.remove(key);

        return res;
    }

    private boolean isInRange(int[][] arr, int row, int col) {
        return row >= 0 && col >= 0 && row < arr.length && col < arr[0].length;
    }
}

File: Data Structures & Algorithms/unique-paths-ii/submission-1.kt
class Solution {

    fun uniquePathsWithObstacles(obstacleGrid: Array<IntArray>): Int {
        if (obstacleGrid[0][0] == 1) return 0
        val arr = Array(obstacleGrid.size) { IntArray(obstacleGrid[0].size) }
        arr[0][0] = 1

        for (row in 0..<arr.size) {
            for (col in 0..<arr[0].size) {
                if (row == 0 && col == 0 || obstacleGrid[row][col] == 1) continue
                val top = if (row - 1 >= 0) arr[row - 1][col] else 0
                val left = if (col - 1 >= 0) arr[row][col - 1] else 0
                arr[row][col] = top + left
            }
        }

        return arr[arr.size - 1][arr[0].size - 1]
    }
}